sys/net/link_layer/ble_ll/ble_ll.c

```diff
@@ -0,0 +1,331 @@
+ */
+static volatile ll_state_t _ll_state = LL_ADVERTISING;
+
+/**
+ * @brief   BLE Advertising address
+ */
+static uint8_t _adv_addr[BLE_ADDR_LEN + 1];
+
+/* Get the BLE Advertising Address of the interface */
+static int _get_adv_address(uint8_t *val, size_t max_len)
+{
+    /* check parameters */
+    if (max_len < BLE_ADDR_LEN) {
+        return -EOVERFLOW;
+    }
+
+    /* get address */
+    val[0] = _adv_addr[0];
+    val[1] = _adv_addr[1];
+    val[2] = _adv_addr[2];
+    val[3] = _adv_addr[3];
+    val[4] = _adv_addr[4];
+    val[5] = _adv_addr[5];
+
+    return BLE_ADDR_LEN;
+}
+
+/* Set the BLE Advertising Address of the interface */
+static int _set_adv_address(uint8_t *val, size_t len)
+{
+    /* check parameters */
+    if (len != BLE_ADDR_LEN) {
+        return -EINVAL;
+    }
+
+    /* set address */
+    _adv_addr[0] = val[5];
+    _adv_addr[1] = val[4];
+    _adv_addr[2] = val[3];
+    _adv_addr[3] = val[2];
+    _adv_addr[4] = val[1];
+    _adv_addr[5] = val[0];
+
+    return BLE_ADDR_LEN;
+}
+
+/* Set ng_netopt parameter value in current BLE interface configuration */
+static int _set(ng_netdev_t *dev, ng_netopt_t opt,
+         void *value, size_t value_len)
+{
+    switch (opt) {
+        case NG_NETOPT_ADV_ADDRESS:
+            return _set_adv_address(value, value_len);
+
+        default:
+            return dev->driver->set(dev, opt, value, value_len);
+    }
+}
+
+/* Get ng_netopt parameter value from current BLE interface configuration */
+static int _get(ng_netdev_t *dev, ng_netopt_t opt,
+         void *value, size_t max_len)
+{
+    switch (opt) {
+        case NG_NETOPT_ADV_ADDRESS:
+            return _get_adv_address(value, max_len);
+
+        default:
+            return dev->driver->get(dev, opt, value, max_len);
+    }
+}
+
+/**
+ * @brief   BLE link layer send function
+ */
+static int _send(ng_netdev_t *dev, ng_pktsnip_t *pkt)
+{
+    uint8_t payload[BLE_MAX_PKT_LEN];
+    ng_pktsnip_t *ble_pkt = NULL;
+    uint8_t ble_pkt_size;
+
+    if (pkt == NULL) {
+        DEBUG("ble_ll: Error handling packet: packet incomplete\n");
+        return -ENOMSG;
+    }
+
+    ble_pkt_size = ng_pkt_len(pkt);
+
+    if (ble_pkt_size > BLE_MAX_PKT_LEN) {
+        ng_pktbuf_release(pkt);
+        DEBUG("ble_ll: Error sending packet: invalid BLE pdu header length\n");
+        return -EOVERFLOW;
+    }
+
+    memcpy(payload, pkt->data, BLE_PDU_HDR_LEN);
+    /* write advertising address to payload */
+    memcpy((payload + BLE_PDU_HDR_LEN), _adv_addr, BLE_ADDR_LEN);
+    ble_pkt_size += BLE_ADDR_LEN;
+
+    /* write optional payload data */
+    if (pkt->next != NULL) {
+        ble_pkt_size += ng_pkt_len(pkt->next);
+        if (ble_pkt_size > BLE_MAX_PKT_LEN) {
+            ng_pktbuf_release(pkt);
+            DEBUG("ble_ll: Error handling packet: payload too large\n");
+            return -EOVERFLOW;
+        }
+        memcpy((payload + BLE_PDU_HDR_LEN + BLE_ADDR_LEN),
+                pkt->next->data, ble_pkt_size);
```

Comment:
Looking at it again though I think I am copying more than I need since `ble_pkt_size` includes headers which have already been copied.